sugar/lib/src/time/zone/providers/base_provider.dart

```diff
@@ -0,0 +1,45 @@
+import 'dart:collection';
+import 'package:sugar/src/time/zone/factory_timezone.dart';
+import 'package:sugar/src/time/zone/timezone.dart';
+
+/// Base class for all timezone providers.
+///
+/// This class is used to provide a Map interface, while using a cache
+/// to store the timezones that have already been fetched.
+/// This class also handles the 'Factory' timezone which is a special timezone
+/// which is treated separately from the other timezones.
+abstract class BaseTimezoneProvider extends UnmodifiableMapBase<String, Timezone> {
+  final _cache = <String, Timezone>{};
+
+  /// Returns a set of all known timezone identifiers.
+  Set<String> listTimezones();
```

Comment:
```suggestion
  Set<String> get timezones;
```

Reply:
What is the difference between this and doing something like BaseTimezoneProvider.keys?

Reply:
`Factory` has to be a key on every timezone provider.
This ensures that every subclass includes the `Factory` key:
```
@override
// Factory is a special timezone that is not in any
// timezone database which is handled separately
late final keys = listTimezones().union({'Factory'});
```

Reply:
We must have `listTimezones()` as another method, or the overridden method would not include `Factory`

Reply:
Done